sound/soc/sof/ipc.c

```diff
@@ -444,8 +457,22 @@ static void ipc_period_elapsed(struct snd_sof_dev *sdev, u32 msg_id)
 	memcpy(&stream->posn, &posn, sizeof(posn));
 
 	/* only inform ALSA for period_wakeup mode */
-	if (!stream->substream->runtime->no_period_wakeup)
-		snd_pcm_period_elapsed(stream->substream);
+	if (!stream->substream->runtime->no_period_wakeup) {
+		if (stream->substream->runtime->status->state ==
+		    SNDRV_PCM_STATE_RUNNING) {
+			snd_pcm_period_elapsed(stream->substream);
```

Comment:
what if we are in DRAINING mode? can't we have a period_elapsed then?

Reply:
so then we exactly want to do the deferred, because alsa will check in period elapsed that if in draining, let's call stop to the dsp. So it will never go up until userspace -> we'll be calling send ipc to dsp in irq handler.